### PR TITLE
[#9] Implement Yes/No reading functions

### PR DESCRIPTION
**Pull request by martinhelmer**
[#9] Implement Yes/No reading functions

remaining work: - DONE
Update CHANGELOG.md. - DONE
move `parseYesNo` to internal module in order to not expose it.  - DONE

### Additional tasks

- [X] Documentation for changes provided/changed
- [X] Tests added
- [x] Updated CHANGELOG.md

### CODE REVIEW

File: src/Iris/Env.hs [hunk 1]
```diff
@@ -82,6 +84,7 @@ mkCliEnv cliEnvSettings@CliEnvSettings{..} = do
             , cliEnvStderrColourMode = stderrColourMode
             , cliEnvAppEnv = cliEnvSettingsAppEnv
             , cliEnvInteractiveMode = interactive
+            , cliEnvInteractiveInputHandle = stdin
```
Comment (martinhelmer) [inline src/Iris/Env.hs:87]: this turned out to not be necessary and will be removed

File: src/Iris/Interactive/Question.hs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        NonInteractive -> pure defaultAnswer
+        Interactive -> liftIO $ ask question >>= loop
+  where
+    loop (Just a) = pure a
+    loop Nothing = ask "  Please answer (Y)es or (N)o " >>= loop
+
+    ask q = do
+        TIO.putStr q
+        hFlush stdout
+        parseYesNo <$> TIO.getLine
```
Comment (chshersh) [inline src/Iris/Interactive/Question.hs:94]: I propose to write the question and the implementation slightly differently

> One thing I was always confused by `(Y)es` and `(N)o` strings earlier is the fact that literal strings `Y(es)` and `(N)o` are not parsed and I actually needed to write `y`

```suggestion
        Interactive -> liftIO loop
  where
    loop :: IO YesNo
    loop = do
        Text.putStr $ q <> " (yes/no)"
        hFlush stdout
        input <- Text.getLine
        case parseYesNo input of
            Just answer -> pure answer
            Nothing -> do
                Text.putStrLn $ "I don't understand your answer: '" <> input <> "'"
                Text.putStrLn "Please, answer yes or no (or y, or n)"
                loop
```

File: src/Iris/Interactive/Question.hs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+@
+app :: App ()
+app = do
+    answer <- Iris.yesno "Would you like to proceed?" Iris.Yes
+    case answer of
+        Iris.Yes -> proceed
+        Iris.No -> Iris.outLn "Aborting"
+
+@
+
```
Comment (chshersh) [inline src/Iris/Interactive/Question.hs:74]: Yes also add an example of what will users see in the output when running this function

File: test/Test/Iris/Interactive/Question.hs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+questionSpec =
+    describe "Question - parse YesNo" $ do
+        describe "should parse to Just Yes" $ do
+            forM_ ["y", "Y", "YES", "yes", "Yes"] $
+                \m -> do
+                    it (unpack m) $ do
+                        pure (parseYesNo m) `shouldReturn` Just Yes
+
+        describe "should parse to Just No" $ do
+            forM_ ["n", "N", "NO", "no", "No"] $
```
Comment (chshersh) [inline test/Test/Iris/Interactive/Question.hs:24]: Let's add one more test case just to be extra sure that all inputs are actually handled 🙂 

```suggestion
            forM_ ["n", "N", "NO", "no", "No", "nO"] $
```

File: test/Test/Iris/Interactive/Question.hs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    -- under test
+    YesNo (..),
+    parseYesNo,
+ )
+
+questionSpec :: Spec
+questionSpec =
+    describe "Question - parse YesNo" $ do
+        describe "should parse to Just Yes" $ do
+            forM_ ["y", "Y", "YES", "yes", "Yes"] $
```
Comment (chshersh) [inline test/Test/Iris/Interactive/Question.hs:18]: Let's test all possible combinations of upper- and lowercase letters.

I can think of the following way to generate a list of all cases we want to handle. Something like this:

```haskell
yesAnswers :: [Text]
yesAnswers = "y" : "Y" : [y <> e <> s | y <- ["y", "Y"], e <- ["e", "E", ""], s <- ["s", "S"]]

File: test/Test/Iris/Interactive/Question.hs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+                \m -> do
+                    it (unpack m) $ do
+                        pure (parseYesNo m) `shouldReturn` Just Yes
+
+        describe "should parse to Just No" $ do
+            forM_ ["n", "N", "NO", "no", "No"] $
+                \m -> do
+                    it (unpack m) $ do
```
Comment (chshersh) [inline test/Test/Iris/Interactive/Question.hs:26]: `unpack` definitely should be imported qualified and used like `Text.unpack` because it's a too common name

File: test/Test/Iris/Interactive/Question.hs (hunk 5 above)
Comment (martinhelmer) [reply test/Test/Iris/Interactive/Question.hs:18]: rewriting the test so all the cases are reported on one row 
```
    Question - parse YesNo
      elementes should parse to Just Yes
        ["y","Y","yes","yeS","yEs","yES","ys","yS","Yes","YeS","YEs","YES","Ys","YS"] [✔]
      elementes should parse to Just No
        ["n","N","NO","no","No","nO"] [✔]
      elementes should parse to Nothing
        ["a","ye","NOone","yesterday","oui"] [✔]
```

a failure becomes:

```
Failures:

  test/Test/Iris/Interactive/Question.hs:34:54:
  1) Iris.Interactive, Question - parse YesNo, elementes should parse to Just No, ["ne","N","NO","no","No","nO"]
       Actual list is not a permutation of expected list!
         expected list contains:   [("ne",Just No), ("N",Just No), ("NO",Just No), ("no",Just No), ("No",Just No), ("nO",Just No)]
         actual list contains:     [("ne",Nothing), ("N",Just No), ("NO",Just No), ("no",Just No), ("No",Just No), ("nO",Just No)]
         the missing elements are: [("ne",Just No)]
         the extra elements are:   [("ne",Nothing)]
```
so it's still easy to see what went wrong in detail.

like?

File: test/Test/Iris/Interactive/Question.hs [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        checkElements ["", "a", "ye", "NOone", "yesterday", "oui"] Nothing
+
+checkElements
+    :: [Text]
+    -> Maybe YesNo
+    -> SpecWith ()
+checkElements values expected = do
+    describe ("elementes should parse to " ++ show expected) $ do
+        it (show values) $ do
+            map (\x -> (x, parseYesNo x)) values `shouldMatchList` map (,expected) values
```
Comment (chshersh) [inline test/Test/Iris/Interactive/Question.hs:29]: I actually liked the previous implementation where each constructor was a separate test-case. It's nice that the testing framework shows proper diffs of lists.

But I find the output with string quotes in lists too noisy and having a single test case per string is actually fine in this particular case because there's a limited number of test cases we want to parse.

So I would rewrite this code like this:
```suggestion
    describe "Question - parse YesNo" $ do
        checkElements yesAnswers (Just Yes)
        checkElements ["n", "N", "NO", "no", "No", "nO"] (Just No)
        checkElements ["", "a", "ye", "NOone", "yesterday", "oui"] Nothing

checkElements
    :: [Text]
    -> Maybe YesNo
    -> SpecWith ()
checkElements values expected = do
    describe ("should parse to " ++ show expected) $
        forM_ values $ \strValue ->
            it (Text.unpack strValue) $
                parseYesNo strValue `shouldBe` expected
```

**Review by chshersh — CHANGES_REQUESTED** [review]
Looks great! Only a few minor suggestions, and it should be good to go 👍🏻

File: CHANGELOG.md [hunk 8]
```diff
@@ -10,6 +10,15 @@ available [on GitHub][2].
...
 <!-- Add new changes here -->
 
+### Added 
+[#9](https://github.com/chshersh/iris/issues/9):
+Implement Yes/No reading functions:
+
+  * Add `yesno` function for asking a question with can be answered with either Yes or No
+  * Add `YesNo` type (`Yes` | `No`)
+
+  (by [@martinhelmer])
```
[the remark this answers is not in the document]
Comment (martinhelmer) [reply CHANGELOG.md:20]: i think the [@martinhelmer] resolves itself to a link if I just add myself to the contributors. (i do not see explicit markdown syntax anywhere else)

[the remark this answers is not in the document]
Comment (chshersh) [reply CHANGELOG.md:20]: Duh, indeed, you're right 🤦🏻 

I used to duplicate the user name in other OSS projects:

* https://github.com/chshersh/tool-sync/blob/da2958384c8e3d0fc2539b7d812d3c3084481788/CHANGELOG.md?plain=1#L16

But while working on `iris` I learned that it indeed could be resolved to itself if the link with this name is present. And then I successfully forgot about what I learned 😅

**Review by chshersh — APPROVED** [review]
All looks great! 👏🏻 

@martinhelmer Could you update the branch on the latest version of `main`? And I'm happy to merge it 🙂
